experimental/sql/src/alg_to_ssa.py

```diff
@@ -0,0 +1,151 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from xdsl.ir import Operation, MLContext, Region, Block
+from typing import List, Type, Optional
+from xdsl.dialects.builtin import ArrayAttr, StringAttr, ModuleOp, IntegerAttr
+
+from xdsl.pattern_rewriter import RewritePattern, GreedyRewritePatternApplier, PatternRewriteWalker, PatternRewriter, op_type_rewrite_pattern
+
+import dialects.rel_alg as RelAlg
+import dialects.rel_ssa as RelSSA
+
+# This file contains the rewrite infrastructure to translate the relational
+# algebra dialect to the relational SSA dialect. The current design has a parent
+# class `RelAlgRewriter` that contains functions used for several `Rewriter`s.
+# All other `Rewriter`s inherit from that class. To introduce SSA, the passes
+# insert the children, that should become operands, before the matched op, then
+# use the `added_operations_before` functionality of the rewriter to link the
+# operands to the new operation.
+
+
+@dataclass
+class RelAlgRewriter(RewritePattern):
+
+  def convert_datatype(self, type_: RelAlg.DataType) -> RelSSA.DataType:
+    if isinstance(type_, RelAlg.String):
+      return RelSSA.String.get(type_.nullable)
+    if isinstance(type_, RelAlg.Int32):
+      return RelSSA.Int32()
+    raise Exception(
+        f"datatype conversion not yet implemented for {type(type_)}")
+
+  def lookup_type_in_schema(self, name: str,
+                            bag: RelSSA.Bag) -> Optional[RelSSA.DataType]:
+    """
+    Looks up the type of name in the schema of bag.
+    """
+    for s in bag.schema.data:
+      if s.elt_name.data == name:
+        return s.elt_type
+    return None
+
+  def lookup_type_in_parent_op(self, name: str, parent_op: Operation):
+    """
+    Crawls through all parent_ops until reaching either a ModuleOp, in which
+    case the lookup failed or reaching an operation with an input bag, that the
+    type can be looked up in.
+    """
```

Comment:
Can you convince me that this always does the right thing?

If I have understood the structure of this dialect correctly, "expressions" are nested inside of "operators," which, in turn, form a program in SSA form. I *guess* that at this point we know the schema of any operator in the top-level program, no? So, from any expression, the first containing *operator* -- not just the first containing `Select` -- should have the relevant schema, no?

If that's the case, I guess that the `Operator` class could/should have a schema property such that you can write this function against the `Operator` interface (with `isinstance(parent_op, Relalg.Operator)` or similar).

Reply:
Yes you are right, as long as we do not have operators with nested expressions that access fields that are not part of the output anymore. I am unsure whether there can be operators like that in relational algebra, I can think of a natural join that, in some sense, removes a column (compared to the cross-product), but obviously that column is part of the result anyway...

My plan was to add matching for more operators, when they are introduced and then look at their inputs, because the name certainly has to be part of one of the input schemas.

Reply:
We haven't really fleshed that out yet but I think that making sure that these names exist is the job of some earlier conversions. In the relational algebra dialects, I think that that should be an invariant: only names can be accessed that are made available by the encompassing operator (with matching types).

Of course, you can spell out what all types do (then add a comment that that's how you plan to extend it) but I think it would be a good thing to factor out that functionality. *Every* operator produces a relation for which we have to be able to say what the schema is and -- as I argue -- the column expression should by design refer to that schema in every operator that is parametrized with an expression...

Reply:
I think the thing I am uncertain about atm is that to me the `column()` expression takes a field of the *input* tuple. AFAIU, what you are suggesting is looking up the type in the *output* relation. Is it really the case that this always works? If so, I think your approach is absolutely valid and makes sense.

Reply:
I implemented your suggestion :)

Reply:
> I think the thing I am uncertain about atm is that to me the `column()` expression takes a field of the _input_ tuple. AFAIU, what you are suggesting is looking up the type in the _output_ relation. Is it really the case that this always works? If so, I think your approach is absolutely valid and makes sense.

I hadn't though about input vs output names. I think I only though about input names as well. One operator that changes the set of output names is the "rename" operator (or similar), which would have to handle the new names with a different mechanism than `column()`.

> I implemented your suggestion :)

Cool, I'll take a look.

Reply:
The other operator that changes the set I can think of is a projection. But both don't actually need expressions to parametrize, so for those two, your approach should work.